### ISSUE TITLE
Safari extension

### ISSUE THREAD

**Issue by boblmartens** [wallabag/wallabagger#214]
I am wondering if anyone has investigated creating a Safari extension so that macOS users can have the same functionality in their browsers.

**Comment by tcitworld** [comment]
cc @bourvill

**Comment by boblmartens** [comment]
An exciting new era may be upon us with the announcements at WWDC20.

**Comment by dmin7b5** [comment]
@boblmartens I tried using their command line converter tool just to see if there would be incompatibilities. And there were many items used in the manifest file that were not supported in Safari. Which makes me think it would take a complete rewrite of this extension.

**Comment by boblmartens** [comment]
I have in mind to maybe take a small stab at it sometime and see how far I can get. I've never done extension development before. Listening to the Zotero list as well, they are looking at the same thing and have found that many of the incompatibilities are tied more directly to things that aren't needed.

I don't know if that is the case here.

**Comment by dmin7b5** [comment]
Here is the list:

Warning: The following keys in your manifest.json are not supported by your current version of Safari. If these are critical to your extension, you should review your code to see if you need to make changes to support Safari:
	description
	name
	version
	default_locale
	commands
	options_ui
	tabs
	activeTab
	scripts
	storage
	manifest_version
	contextMenus
	browser_action
	icons

**Comment by boblmartens** [comment]
I should also state that I built from the installed Chrome extension and not from the source itself.

**Comment by boblmartens** [comment]
Also, it looks like Apple has a tutorial posted as well: https://developer.apple.com/news/?id=qiz0arxc&

**Comment by jandamm** [comment]
@boblmartens do you have pushed the code somewhere?

**Comment by boblmartens** [comment]
I do not have the code pushed at all. What would be the best way to do it? Should I create my own repo?

**Comment by jandamm** [comment]
I'm not a 100% sure.
I wanted to start a Mac app which can act both as a share sheet and a Safari extension.
But the Safari Extension would be 90% I need so if there would be a working one I wouldn't start the project.
But I probably wouldn't use the new chrome extension api but create a native extension since I'm more familiar with it.

If you think it could be integrated here (sharing the extension code) I think forking would be a good idea. Otherwise I'd create a new repo.

**Comment by boblmartens** [comment]
I will try and set aside some time in the coming week to play around with it. It was super rough, but it did "work" for the most part. I will report back here when I have some code for everyone to look at along with what I did to get it working to see if we can reproduce it and then start working on the bugs.

**Comment by boblmartens** [comment]
@jandamm I should probably also add that this extension would be missing a key component ... the ability to share from other applications. Just a thought.

**Comment by jandamm** [comment]
@boblmartens I think I'll go ahead and create a native Share and Safari Extension (they can share about 90% of the code).
I probably will only add a confirmation, no way to tag items since I wouldn't use it. (Maybe I'll add a way to set a default tag, which allows you to filter the newly and untagged items - not sure yet)

**Comment by boblmartens** [comment]
@jandamm That sounds great. I may still go the route I outlined so that we have a copy of the Chrome extension for those who want it.

If I can pull it off.

If I can find time.

Gotta love it.

**Comment by jandamm** [comment]
Actually we could bundle up both extensions in one wrapper app.
One as a "Speed Save" the other as a default one.

I think it would be critical that your extension either shares the source code with the one for Chrome or have some kind of automatic migration. Otherwise they might get out of sync quickly.

**Comment by boblmartens** [comment]
Agreed, I'd want to share the code from the Chrome extension as much as possible and then I'd need to keep up with the changes on one.

I'll maybe fork this into my account and see what I can do to get it working more smoothly for Safari.

**Comment by jandamm** [comment]
Just had a look what the chrome extension provides. This should be pretty easy to replicate (just the tag completion would take a while).
Maybe I'll add a Speed Save mode which you can toggle. Otherwise a UI opens where you can set the entry `starred`, `archived` or add tags.

**Comment by boblmartens** [comment]
If there is any help I can render (testing maybe) let me know.

**Comment by jandamm** [comment]
@boblmartens if you don't mind to install Xcode you can check out the Safari Extension here:
https://github.com/jandamm/wallabag-macos

Todos:
- App Icon and other icons 
- Confirmation/Error UI when adding an URL
- Share Extension

**Comment by jandamm** [comment]
@boblmartens if/when you've gotten around to testing it please give feedback here (if everything went well) or create an issue in the other repo if you spotted a bug.

**Comment by boblmartens** [comment]
After I was able to get it built, I was able to get it to successfully save an article. I'll create some issues in the repo so that we can have more discussion over there.

**Comment by navanchauhan** [comment]
![image](https://user-images.githubusercontent.com/29234112/107959497-84ad6d80-6fc9-11eb-9f5a-e4be26a6eb49.png)

I converted the 1.10.4 binary (crx) and it is working flawlessly. 

Steps for others to reproduce:

```bash
wget https://github.com/wallabag/wallabagger/raw/master/wallabagger.crx
unzip wallabagger.crx
xcrun safari-web-extension-converter .
```

I believe jandamm did the same and am going to try to contribute to his repository rather than going off alone. Just wanted to leave the steps. It compiles and runs without any errors, so I believe Apple have updated Safari with more capabilities

**Comment by jandamm** [comment]
@navanchauhan nope, I did not do this.
I've created an `Safari App Extension` written in native code. Thus being able to further evolve the app by adding more extensions. In theory a Wallabag macOS app with native reader would be possible while still sharing the same authentication, storage and api code.

You've converted the existing Chrome extension into a `Safari Web Extension` which was recently introduced by Apple and allows to share JS code written for Chrome, Firefox and other browsers with Safari.

Nice that it you found a way to convert the extension 👍

**Comment by dmin7b5** [comment]
@navanchauhan I tried your method and it worked for me too. One additional step needed. You have to go to "Develop" menu in Safari and check "Allow unsigned extensions". Once I did that I could finish installing the extension, enter my wallabag credentials and I was good to go.

**Comment by navanchauhan** [comment]
@jandamm huh, what should we do then? Should we ping the owners of the Wallabag dev account to add the conversion to their build pipeline so they can publish it with their credentials?

@dmin7b5 Oh yes, I forgot about that 😃

For anyone else stumbling upon this issue, I have built and attached the extension app

[Wallabagger.app.zip](https://github.com/wallabag/wallabagger/files/5982877/Wallabagger.app.zip)

**Comment by tcitworld** [comment]
Sorry to bump into this so late, but FYI we have [an iOS app](https://github.com/wallabag/ios-app) that I think @bourvill wanted to work on macOS as well.
Even if two separate apps coexist, it would be nice to make use of a single API wrapper at least: https://github.com/wallabag/WallabagKit

**Comment by AizatHawari** [comment]
> > Hi, I'm a big fan of Wallabag since I've tried it 2 months ago. Unfortunately the extension of the app posted by @navanchauhan is no longer working with Safari Version 15.0 (16612.1.29.41.4, 16612), at least to me.
> > Does anyone has any updates on the extension that works with Safari? Did it work for you?
> 
> It works for me. You have to compile it with the latest XCode. I'm running macOS Big Sur 11.6 with Safari Version 15.0 (16612.1.29.41.4, 16612) with no problems. The steps I use are:
> 
> Make sure to check "Show Develop menu in menu bar" under the "Advanced" tab under Safari preferences. Click on the "Develop" menu and check "Allow Unsigned Extensions" way at the bottom. Compile the app in XCode targeting your local Mac. The extension will appear after the build. Open the Wallabag preference window and fill in your credentials. Enjoy this great extension

Thank you so much, your method works! The "Allow Unsigned Extensions" worked for me, I didn't have to compile the app in XCode.
Just out of curiosity, the "token expires in 59 minutes", does it mean that after 1 hour, I would need to manually grant access again?

**Comment by AizatHawari** [comment]
> For anyone else stumbling upon this issue, I have built and attached the extension app
> 
> [Wallabagger.app.zip](https://github.com/wallabag/wallabagger/files/5982877/Wallabagger.app.zip)

Hi, anyone was able to package an extension for Safari? This is not in my area at all, not sure where to start. Any help would be appreciated.

**Comment by aniqueta** [comment]
It's not a FOSS option, but [ReadKit 3](https://readkit.app) that just came out yesterday has a "Read Later in ReadKit" share sheet extension that works across iOS/iPadOS and macOS and supports Wallabag. I would also argue that ReadKit is the best Wallabag client across these platforms anyway.

**Comment by jandamm** [comment]
> I've created an `Safari App Extension` written in native code. Thus being able to further evolve the app by adding more extensions. In theory a Wallabag macOS app with native reader would be possible while still sharing the same authentication, storage and api code.

In the meantime I've created an Apple Developer Account and uploaded the [app](https://apps.apple.com/de/app/wallabag-quicksave/id1621482657?l=en&mt=12).

You can also have a look at [the source code](https://github.com/jandamm/wallabag-macos).
